File: placementprepration/tree/topview.java
import java.io.*;
import java.util.*;
import org.w3c.dom.Node;
class TreeNode{
    int val;
    TreeNode left, right;
    TreeNode(int val){
        this.val=val;
        this.right=null;
        this.left=null;
    }
}
public class topview{
    public class Pair{
        TreeNode node;
        int line;
        Pair(TreeNode node, int line){
            this.node=node;
            this.line=line;
        }
    }
    public List<Integer> Topviewtree(TreeNode root){
        List<Integer>ans=new ArrayList<>();
        if(root==null){
            return ans;
        }
        Map<Integer, Integer>mpp=new TreeMap<>();
        Queue<Pair>q=new LinkedList<>();
        q.add(new Pair(root,0));
        while(!q.isEmpty()){
            Pair temp=q.poll();
            TreeNode node=temp.node;
            int line=temp.line;
            if(!mpp.containsKey(line)){
                mpp.put(line,node.val);
            }
            if(node.left!=null){
                q.add(new Pair(node.left,line-1));
            }
            if(node.right!=null){
                q.add(new Pair(node.right,line+1));
            }
        }
        for(int it:mpp.values()){
            ans.add(it);
        }
        return ans;

    }
}

File: placementprepration/tree/topview.py
from collections import deque


class Node:
    def __init__(self, val):
        self.val = val
        self.left = None
        self.right = None


def top_view(root):
    if not root:
        return []
    hd_map = {}
    q = deque([(root, 0)])
    while q:
        node, hd = q.popleft()
        if hd not in hd_map:
            hd_map[hd] = node.val
        if node.left:
            q.append((node.left, hd - 1))
        if node.right:
            q.append((node.right, hd + 1))
    return [hd_map[k] for k in sorted(hd_map.keys())]


if __name__ == "__main__":
    r = Node(1)
    r.left = Node(2)
    r.right = Node(3)
    r.left.right = Node(4)
    r.left.right.right = Node(5)
    r.left.right.right.right = Node(6)
    print(top_view(r))
